cooked-validators/src/Cooked/Attack/AddConstraints.hs

```diff
@@ -0,0 +1,264 @@
+          -- not. Since the current implementation of the 'Before' and 'After'
+          -- constraints includes the endpoints, I will do so here as well.
+          Before b -> Pl.intersection (Pl.to b)
+          After a -> Pl.intersection (Pl.from a)
+          ValidateIn i -> Pl.intersection i
+          _ -> id
+      )
+      Pl.always
+      timeConstraints
+
+-- | This attack restricts the validity time range of a transaction to the
+-- intersection of the given range and its current validity time range.
+--
+-- It also "cleans up" the time constraints (i.e. 'Before', 'After', and
+-- 'ValidateIn'), so that the resulting transaction has only one 'ValidateIn'
+-- constraint.
+--
+-- Returns the validity time range of the unmodified transaction.
+addValidateInAttack :: L.POSIXTimeRange -> Attack L.POSIXTimeRange
+addValidateInAttack range = do
+  oldRange <- removeTimeConstraintsAttack
+  overAttack miscConstraintsL (ValidateIn (range `Pl.intersection` oldRange) :)
+  return oldRange
+
+-- | This attack removes all 'SignedBy' constraints from a transaction. It
+-- returns a list of the signers it removed.
+removeSignedByAttack :: Attack [L.PubKeyHash]
+removeSignedByAttack = do
+  removed <- removeMiscConstraintsAttack (\case SignedBy _ -> True; _ -> False)
+  return $ concatMap (\case SignedBy s -> s; _ -> []) removed
+
+-- | This attack adds signers to a transaction. It also combines all 'SignedBy'
+-- constraints of the transaction into one.
+--
+-- Returns a list of all signers that were added (i.e. that were not already
+-- present on the unmodified transaction).
+addSignedByAttack :: [L.PubKeyHash] -> Attack [L.PubKeyHash]
+addSignedByAttack signers = do
+  oldSigners <- removeSignedByAttack
+  let newSigners = filter (`notElem` oldSigners) signers
+  overAttack miscConstraintsL (SignedBy (newSigners ++ oldSigners) :)
+  return newSigners
+
+-- * Adding and Removing 'OutConstraint's
+
+-- | Add an 'OutConstraint' to a transaction. The additional constraint will be
+-- added to the end of the list of output constraints. This is because some
+-- contracts rely on the shape of (the initial segment of) the list of
+-- transaction outputs. Try 'permutOutAttack' if you want to change the ordering
+-- of output constraints.
+addOutConstraintAttack :: OutConstraint -> Attack ()
+addOutConstraintAttack oc = overAttack outConstraintsL (++ [oc])
+
+-- | Add a 'paysPK' constraint to a transaction. The additional constraint will
+-- be added to the end of the list of output constraints.
+addPaysPKAttack ::
+  L.PubKeyHash ->
+  L.Value ->
+  Attack ()
+addPaysPKAttack h v = addOutConstraintAttack $ paysPK h v
+
+-- | Add a 'PaysScript' to a transaction. The additional constraint will be
+-- added to the end of the list of output constraints.
+addPaysScriptAttack ::
+  PaysScriptConstrs a =>
+  L.TypedValidator a ->
+  L.DatumType a ->
+  L.Value ->
+  Attack ()
+addPaysScriptAttack v d x = addOutConstraintAttack $ PaysScript v d x
```

Comment:
Aren't `addPaysScriptAttack` and `addPaysPKAttack` rather redundant with `addOutConstraintAttack`?